### PR TITLE
Another strategy to avoid local minima

### PR DESCRIPTION
**Pull request by abstractingagent**
I discovered another strategy in order for one to be able to train a neural ODE and avoid local minima. I used flux because I didn't know how to do it with sciml, but it's pretty straightforward. Simply train both the initial conditions and the parameters first, and then reset the IC back to it's original values and train only the parameters. Follow it up with a longer ADAM run or BFGS for fine tuning. Sometimes in the first training round it doesn't fully converge (for the longer timespan like 0,10), but just run both iterations until their end or add more iterations and it always converges. Going to post the code here and then update the PR with the documentation added once the idea is verified.

```julia
using DiffEqFlux, OrdinaryDiffEq, Flux, Optim, Plots, DifferentialEquations


#Starting with an example with tspan (0, 5)
u0 = Float32[2.0; 0.0]
datasize = 30
tspan = (0.0f0, 5.0f0)
tsteps = range(tspan[1], tspan[2], length = datasize)

function trueODEfunc(du, u, p, t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end

prob_trueode = ODEProblem(trueODEfunc, u0, tspan)
ode_data = Array(solve(prob_trueode, Tsit5(), saveat = tsteps))

dudt2 = Chain(Dense(2,16, tanh),
             Dense(16,2))


p,re = Flux.destructure(dudt2) # use this p as the initial condition!
dudt(u,p,t) = re(p)(u) # need to restrcture for backprop!
prob = ODEProblem(dudt,u0,tspan)

function predict_n_ode()
    Array(solve(prob,u0=u0,p=p, saveat=tsteps, maxiters = 1e10))
end

function loss_n_ode()
      pred = predict_n_ode()
      sqnorm(x) = sum(abs2, x)
      loss = sum(abs2,ode_data .- pred)
      loss
end

function cb(;doplot=false) #callback function to observe training
    pred = predict_n_ode()
    display(sum(abs2,ode_data .- pred))
    if doplot
      # plot current prediction against data
      pl = plot(tsteps,ode_data[1,:],label="data")
      plot!(pl,tsteps,pred[1,:],label="prediction")
      display(plot(pl))
    end
    return false
end
predict_n_ode()
loss_n_ode()
cb(;doplot=true)

data = Iterators.repeated((), 1000)

Flux.train!(loss_n_ode, Flux.params(u0, p), data,
                    Flux.Optimise.ADAM(0.05), cb = cb)

u0 = Float32[2.0; 0.0]
Flux.train!(loss_n_ode, Flux.params(p), data,
            Flux.Optimise.ADAM(0.05), cb = cb)

#Now use the same technique for longer tspan (0, 10)
datasize = 30
tspan = (0.0f0, 10.0f0)
tsteps = range(tspan[1], tspan[2], length = datasize)

prob_trueode = ODEProblem(trueODEfunc, u0, tspan)
ode_data = Array(solve(prob_trueode, Tsit5(), saveat = tsteps))

dudt2 = Chain(Dense(2,16, tanh),
             Dense(16,2))


p,re = Flux.destructure(dudt2) # use this p as the initial condition!
dudt(u,p,t) = re(p)(u) # need to restrcture for backprop!
prob = ODEProblem(dudt,u0,tspan)
predict_n_ode()
loss_n_ode()
cb(;doplot=true)

data = Iterators.repeated((), 1500)

Flux.train!(loss_n_ode, Flux.params(u0, p), data,
                    Flux.Optimise.ADAM(0.05), cb = cb)

u0 = Float32[2.0; 0.0]
Flux.train!(loss_n_ode, Flux.params(p), data,
            Flux.Optimise.ADAM(0.05), cb = cb)
```

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
I don't think you included the example you intended to? Instead it's just a single `.`?

**Comment by abstractingagent** [comment]
I included the example in the description for you to try out, and then I was going to update the pull request with it once you verified?

**Comment by ChrisRackauckas** [comment]
Yeah it checks out.

**Comment by abstractingagent** [comment]
Awesome, going to update the PR - before I do though, what would the equivalent version of this be for scimltrain? Could add that as well.

**Comment by ChrisRackauckas** [comment]
You'd have to have two loss functions: one with the extended vector and one without. It's okay though: we can add that extension later. It's more about capturing the idea.

**Comment by abstractingagent** [comment]
Updated it! Let me know if you need me to do anything else.

**Comment by ChrisRackauckas** [comment]
Great thanks!
